src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -703,51 +808,42 @@ private async Task SendDataAsync(ReadOnlyMemory<byte> buffer, CancellationToken
             {
                 ReadOnlyMemory<byte> remaining = buffer;
 
-                while (remaining.Length > 0)
+                // Deal with ActiveIssue #9071:
+                // Custom HttpContent classes do not get passed the cancellationToken.
+                // So, inject the expected CancellationToken here, to ensure we can cancel the request body send if needed.
+                CancellationTokenSource customCancellationSource = null;
+                if (!cancellationToken.CanBeCanceled)
                 {
-                    int sendSize = await _streamWindow.RequestCreditAsync(remaining.Length, cancellationToken).ConfigureAwait(false);
-
-                    ReadOnlyMemory<byte> current;
-                    (current, remaining) = SplitBuffer(remaining, sendSize);
-
-                    await _connection.SendStreamDataAsync(_streamId, current, cancellationToken).ConfigureAwait(false);
+                    cancellationToken = _requestBodyCancellationSource.Token;
+                }
+                else if (cancellationToken != _requestBodyCancellationSource.Token)
+                {
+                    // User passed a custom CancellationToken.
+                    // We can't tell if it includes our Token or not, so assume it doesn't.
+                    customCancellationSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _requestBodyCancellationSource.Token);
+                    cancellationToken = customCancellationSource.Token;
                 }
-            }
 
-            public void Dispose()
-            {
-                lock (SyncObject)
+                using (customCancellationSource)
                 {
-                    if (!_disposed)
+                    while (remaining.Length > 0)
                     {
-                        _disposed = true;
+                        int sendSize = await _streamWindow.RequestCreditAsync(remaining.Length, cancellationToken).ConfigureAwait(false);
 
-                        _streamWindow.Dispose(_abortException);
-                        _responseBuffer.Dispose();
+                        ReadOnlyMemory<byte> current;
+                        (current, remaining) = SplitBuffer(remaining, sendSize);
 
-                        // TODO: ISSUE 31310: If the stream is not complete, we should send RST_STREAM
+                        await _connection.SendStreamDataAsync(_streamId, current, cancellationToken).ConfigureAwait(false);
```

Comment:
What happens if an exception occurs on this path?  It'll propagate out to the Stream.WriteAsync caller, which presumably is the correct behavior.  But will such errors also be noticed by the rest of the request/response processing and cause the request to be canceled?  Or is it possible we could end up hanging because the caller will stop writing more data but the rest of the request processing thinks it'll still be sending more?

Reply:
The idea is that the exception will propagate out to SendRequestBodyAsync, which will then catch it and cancel everything.

It's possible that a badly behaved HttpContent could eat this exception, and then continue on somehow... perhaps try to write again, or just complete like it succeeded. Possibly we could get into a bad state if that happens. I'm not sure whether this is worth dealing with or not... thoughts?